open_city_profile/graphene.py

```diff
@@ -178,3 +183,55 @@ def __init_subclass_with_meta__(
             _meta=_meta,
             **options,
         )
+
+
+def send_allowed_data_field_sentry_warning(
+    message: str, service: Optional[Service], field_name: str
+):
+    with sentry_sdk.push_scope() as scope:
+        scope.set_extra("service", getattr(service, "name", "Service not identified"))
+        scope.set_extra("field_name", field_name)
+
+        sentry_sdk.capture_message(message, level="warning")
+
+
+class AllowedDataFieldsMiddleware:
+
+    def _is_profile_query(self, info):
+        return any(
+            selection.name.value in ("myProfile", "profile", "profiles")
+            for selection in info.operation.selection_set.selections
+        )
+
+    def resolve(self, next, root, info, **args):
+        try:
+            is_profile_query = self._is_profile_query(info)
+            node = info.parent_type.graphene_type
+        except AttributeError:
+            is_profile_query = False
+            node = None
+
+        if is_profile_query and getattr(node, "check_allowed_data_fields", False):
+            field_name = to_snake_case(getattr(info, "field_name", ""))
+
+            if not getattr(info.context, "service", False):
+                if settings.ENABLE_ALLOWED_DATA_FIELDS_RESTRICTION:
+                    raise ServiceNotIdentifiedError("Service not identified")
+
+                send_allowed_data_field_sentry_warning(
+                    "Allowed data field exception would occur: Service not identified",
+                    None,
+                    field_name,
+                )
+
+            if not node.is_field_allowed_for_service(field_name, info.context.service):
+                if settings.ENABLE_ALLOWED_DATA_FIELDS_RESTRICTION:
+                    raise FieldNotAllowedError("Field is not allowed for service.")
+
+                send_allowed_data_field_sentry_warning(
+                    "Allowed data field exception would occur. Field is not allowed for service.",
+                    info.context.service,
+                    field_name,
+                )
+
+        return next(root, info, **args)
```

Comment:
Just thinking here. What if `AllowedDataFieldsMixin` was used on the models instead and not on the graphene node classes (so `Profile` and `VerifiedPersonalInformation`).

You could then do something like:

```suggestion
    def resolve(self, next, root, info, **kwargs):
        if getattr(root, "check_allowed_data_fields", False):
            field_name = to_snake_case(info.field_name)

            if not getattr(info.context, "service", False):
                if settings.ENABLE_ALLOWED_DATA_FIELDS_RESTRICTION:
                    raise ServiceNotIdentifiedError("Service not identified")
                
                logger.warning(f"Allowed data field exception would occur: Service not identified")

            if not root.is_field_allowed_for_service(field_name, info.context.service):
                if settings.ENABLE_ALLOWED_DATA_FIELDS_RESTRICTION:
                    raise FieldNotAllowedError("Field is not allowed for service.")

                logger.warning(f"Allowed data field exception would occur. Field '{field_name}' is not allowed for service {info.context.service}.")

        return next(root, info, **kwargs)
```

Reply:
This is true and I actually started about going for the model side. But then I thought since the restriction concerns the api, the restriction logic should be in the api not in the model 🤔 
But I'm open to this way of solution also if there are strong opinions towards this.

Reply:
Don't know if it's good or not, but the solution which I suggested wouldn't need `_is_profile_query` method at all. It would apply anytime a model has `check_allowed_data_fields` set to `True` i.e. not just on those specific queries.

Reply:
I changed those checks being in the model and thus be using the root in the middleware 👌